src/main/java/net/spy/memcached/protocol/TCPMemcachedNodeImpl.java

```diff
@@ -637,15 +637,22 @@ public final void setupForAuth(String cause) {
   }
 
   public final void shutdown() throws IOException {
-    if (channel != null) {
-      channel.close();
-      sk = null;
-      if (toWrite > 0) {
-        getLogger().warn(
-                "Shut down with %d bytes remaining to write",
-                toWrite);
+    try {
+      if (channel != null) {
+        channel.close();
+        if (toWrite > 0) {
+          getLogger().warn(
+              "Shut down with %d bytes remaining to write",
+              toWrite);
+        }
+        getLogger().debug("Shut down channel %s", channel);
+      }
+    } finally {
+      if (sk != null) {
+        sk.attach(null);
+        sk.cancel();
```

Comment:
channel close 면, selector에서 등록된 사실이 cancel된다고 들었습니다.
그래도, close 하는 과정에서 문제가 있을 수 있으니, 명시적으로 cancel하는 것이 안전하여 위와 같이 구현해 둔 거죠?

그리고, cancel 먼저 호출하여 selector에서 sk's channel 등록을 취소한 후에, attach(null) 호출하는 것이 순서상 맞지 않는지 ?

Reply:
@jhpark816 

implCloseSelectionChannel()이 실패하면 selection key가 cancel되지 않습니다. 결국은 selection key를 같이 cancel해주는 것이 안전합니다.

```java
protected final void implCloseChannel() throws IOException {
        implCloseSelectableChannel();
        synchronized (keyLock) {
            int count = (keys == null) ? 0 : keys.length;
            for (int i = 0; i < count; i++) {
                SelectionKey k = keys[i];
                if (k != null)
                    k.cancel();
            }
        }
    }
```

```java
  protected void implCloseSelectableChannel() throws IOException {
    synchronized(this.stateLock) {
      this.isInputOpen = false;
      this.isOutputOpen = false;
      if (this.state != 4) {
        nd.preClose(this.fd);
      }

      if (this.readerThread != 0L) {
        NativeThread.signal(this.readerThread);
      }

      if (this.writerThread != 0L) {
        NativeThread.signal(this.writerThread);
      }

      if (!this.isRegistered()) {
        this.kill();
      }

    }
  }
```


sk.attach(null), sk.cancel() 호출 순서는 상관 없습니다. cancel시 내부의 attachment 객체를 참조하지 않습니다.

```java
private volatile Object attachment = null;

    private static final AtomicReferenceFieldUpdater<SelectionKey,Object>
        attachmentUpdater = AtomicReferenceFieldUpdater.newUpdater(
            SelectionKey.class, Object.class, "attachment"
        );

    /**
     * Attaches the given object to this key.
     *
     * <p> An attached object may later be retrieved via the {@link #attachment()
     * attachment} method.  Only one object may be attached at a time; invoking
     * this method causes any previous attachment to be discarded.  The current
     * attachment may be discarded by attaching <tt>null</tt>.  </p>
     *
     * @param  ob
     *         The object to be attached; may be <tt>null</tt>
     *
     * @return  The previously-attached object, if any,
     *          otherwise <tt>null</tt>
     */
    public final Object attach(Object ob) {
        return attachmentUpdater.getAndSet(this, ob);
    }

    /**
     * Retrieves the current attachment.
     *
     * @return  The object currently attached to this key,
     *          or <tt>null</tt> if there is no attachment
     */
    public final Object attachment() {
        return attachment;
    }
```

```java
    public final void cancel() {
        // Synchronizing "this" to prevent this key from getting canceled
        // multiple times by different threads, which might cause race
        // condition between selector's select() and channel's close().
        synchronized (this) {
            if (valid) {
                valid = false;
                ((AbstractSelector)selector()).cancel(this);
            }
        }
    }
```

```java
    void cancel(SelectionKey k) {                       // package-private
        synchronized (cancelledKeys) {
            cancelledKeys.add(k);
        }
    }
```

Reply:
@hjyun328 
- `attach(NULL);`만 수행하고, `cancel()` 하지 않은 상태라면,
   selection key가 selector에 등록된 상태이지만 attachment는 null인 상태가 됩니다.
- 만약, select() 작업을 다른 thread가 수행한다고 가정한다면,
   - select() 내부 동작에서 attachment가 null인 selection key를 만나게 됩니다.
   - 아마 해당 seleciton key를 처리하지 않고 무시할 것으로 예상되어 문제는 없을 것 같습니다.
   - 또한, arcus-java-client는 1개 IO thread가 select() 작업과 node shutdown 작업을 모두 수행하므로,
      본 가정 또한 성립되지 않는 환경입니다.
- 단지, selection key가 selector에서 cancel()되기 전까지 valid attachment를 가지는 것이 
  의미적으로 맞을 것 같다는 의견입니다.

Reply:
@hjyun328 
다시 보니, selectionkey cancel(), attach(NULL) 이후에 channel close 순서가 맞지 않나 생각됩니다.
생성할 때를 보면, channel 생성하고 connect() 한 후에,
selectionkey에 attach하여 selectionkey 생성하는 순서이므로, 그 역순이 적합하지 않나 생각됩니다.

Reply:
@jhpark816 
추가로 유저에 의해서 shutdown하는 것과 
내부에서 shutdown 하는 것을 구분해야 할 것 같습니다.

유저에 의해 shutdown이 호출되면, I/O 쓰레드가 돌아가고 있는 도중에 null 객체를 레퍼런스 할 수 있기 때문입니다.

Reply:
@hjyun328 
closeChannel()이 어떤지?

Reply:
@jhpark816 
괜찮은 것 같습니다. shutdown과 별도로 내부에서만 close하는 메소드를 분리해서 수정하겠습니다.